td-exception/src/idt.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+//! Manipulate x86_64 Interrupt Descriptor Table (IDT).
+//!
+//! Setup a stub IDT for td-shim, which assumes 1:1 mapping between physical address and virtual
+//! address in identity mapping mode.
+//!
+//! It also handles Virtualization Interrupt for Intel TDX technology.
+
+use core::mem::{self, size_of};
+use core::slice::from_raw_parts_mut;
+
+use bitflags::bitflags;
+use lazy_static::lazy_static;
+
+use crate::interrupt;
+
+extern "win64" {
+    fn sidt_call(addr: usize);
+    fn lidt_call(addr: usize);
+    fn read_cs_call() -> u16;
+}
+
+lazy_static! {
+    static ref INIT_IDT: Idt = Idt::new();
+}
+
+#[repr(C, packed)]
+pub struct DescriptorTablePointer {
+    limit: u16,
+    base: u64,
+}
+
+#[no_mangle]
+/// # Safety
+///
+/// This function is unsafe because of the lidt_call()
+pub unsafe fn init() {
+    let mut idtr = DescriptorTablePointer { limit: 1, base: 0 };
+    let current_idt = &INIT_IDT.entries;
+
+    idtr.limit = (current_idt.len() * mem::size_of::<IdtEntry>() - 1) as u16;
+    idtr.base = current_idt.as_ptr() as u64;
+
+    lidt_call(&idtr as *const DescriptorTablePointer as usize);
+}
+
+pub type IdtEntries = [IdtEntry; 256];
+
+// 8 alignment required
+#[repr(C, align(8))]
+pub struct Idt {
+    pub entries: IdtEntries,
+}
+
+impl Default for Idt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Idt {
+    pub fn new() -> Self {
+        let mut idt = Self {
+            entries: [IdtEntry::new(); 256],
+        };
+        idt.init();
+        idt
+    }
+
+    pub fn init(&mut self) {
+        let current_idt = &mut self.entries;
+
+        // Set up exceptions
+        current_idt[0].set_func(interrupt::divide_by_zero);
+        current_idt[1].set_func(interrupt::debug);
+        current_idt[2].set_func(interrupt::non_maskable);
+        current_idt[3].set_func(interrupt::breakpoint);
+        current_idt[4].set_func(interrupt::overflow);
+        current_idt[5].set_func(interrupt::bound_range);
+        current_idt[6].set_func(interrupt::invalid_opcode);
+        current_idt[7].set_func(interrupt::device_not_available);
+        current_idt[8].set_func(interrupt::double_fault);
+        // 9 no longer available
```

Comment:
This is reserved by Intel CPU, according to Software Development Manual Volume 3, 6.15 - Exception and Interrupt Reference.

I would like to understand in which case this may be called.
1) If it is hardware, that means, we dont trust hardware.
However, if we dont trust hardware, then how we believe the hardware will follow SDM to run into this handler?

2) If it is software, that means, we dont trust software.
Then we need setup all 256 interrupt handlers. I think it is burden and not worth to do it.

Reply:
I think so.

If we think software bug may trigger 9, then it may trigger any handler. That means, we need setup handler for all 256 handlers.